### PR TITLE
unit test(operator): Add comprehensive unit tests for trustee module

### PR DESCRIPTION
**Pull request by 6-dehan**
This commit introduces a full suite of 15 unit tests for the functions within `trustee.rs`, validating pure logic, error handling, and both simple and complex Kubernetes API interactions using a mocked client.

-   `test_get_image_pcrs_success`: Verifies that a valid JSON string in a ConfigMap is correctly deserialized into the `ImagePcrs` struct.
-   `test_get_image_pcrs_no_data`: Ensures an error is returned when the ConfigMap's `data` field is missing.
-   `test_get_image_pcrs_invalid_json`: Confirms that an error is propagated when the data contains an invalid JSON string.
-   `test_generate_luks_key_returns_correct_size`: A sanity check to validate that `generate_luks_key` runs without errors and returns a key of the expected 32-byte length.

These tests validate the idempotency and error handling of functions that perform a single `create` operation, primarily testing the `info_if_exists!` macro.

-   `test_create_rv_config_map_success`: Verifies the function returns `Ok(())` on a successful API response (200 OK).
-   `test_create_rv_config_map_already_exists`: Verifies the function correctly handles a 409 Conflict and returns `Ok(())`, confirming idempotency.
-   `test_create_rv_config_map_generic_error`: Ensures a generic API error (e.g., 500) is properly propagated as an `Err`.
-   `test_generate_resource_policy_success`: Validates the success path for the `generate_resource_policy` function.
-   `test_generate_kbs_https_certificate_success`: Validates the success path for the `generate_kbs_https_certificate` function.
-   `test_generate_kbs_configurations_success`: Validates the success path for the `generate_kbs_configurations` function.
-   `test_generate_attestation_policy_success`: Validates the success path for the `generate_attestation_policy` function.
-   `test_generate_kbs_success`: Validates the success path for the `generate_kbs` function.

These tests use a stateful mock client to simulate entire operational flows involving multiple API calls.

-   `test_recompute_reference_values_flow`: Verifies the complete `GET (PCRs) -> GET (RV map) -> PUT (RV map)` sequence executes successfully.
-   `test_generate_secret_flow_success`: Validates the full `CREATE (Secret) -> GET (KbsConfig) -> PATCH (KbsConfig)` workflow for adding a new secret.
-   `test_generate_secret_already_present_in_spec`: Tests the boundary condition where a secret ID already exists in the KbsConfig spec, ensuring the function exits early without making a redundant PATCH call.

### CODE REVIEW

**Comment by 6-dehan** [comment]
Hi @Jakob-Naucke, could you please help review this when you're free?   
```

$ cargo test -p operator
warning: unused import: `chrono::Utc`
 --> operator/src/reference_values.rs:2:5
  |
2 | use chrono::Utc;
  |     ^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: `operator` (bin "operator" test) generated 1 warning (run `cargo fix --bin "operator" --tests` to apply 1 suggestion)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.60s
     Running unittests src/main.rs (target/debug/deps/operator-b20e9e14510aee9a)

running 15 tests
test trustee::tests::test_create_rv_config_map_already_exists ... ok
test trustee::tests::test_create_rv_config_map_generic_error ... ok
test trustee::tests::test_create_rv_config_map_success ... ok
test trustee::tests::test_generate_attestation_policy_success ... ok
test trustee::tests::test_generate_kbs_success ... ok
test trustee::tests::test_generate_kbs_https_certificate_success ... ok
test trustee::tests::test_generate_kbs_configurations_success ... ok
test trustee::tests::test_get_image_pcrs_no_data ... ok
test trustee::tests::test_get_image_pcrs_invalid_json ... ok
test trustee::tests::test_get_image_pcrs_success ... ok
test trustee::tests::test_generate_resource_policy_success ... ok
test trustee::tests::test_recompute_reference_values_flow ... ok
test trustee::tests::test_generate_luks_key_returns_correct_size ... ok
test trustee::tests::test_generate_secret_already_present_in_spec ... ok
test trustee::tests::test_generate_secret_flow_success ... ok

test result: ok. 15 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
```

**Comment by Jakob-Naucke** [comment]
> test result: ok. 15 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

on which revision did you have these test successes? the oldest one gives me
```
error[E0277]: the trait bound `KbsConfigSpec: Default` is not satisfied
   --> operator/src/trustee.rs:795:27
    |
795 |                         ..Default::default()
    |                           ^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `KbsConfigSpec`
```
and I think the added line in the newest revision in `crds` isn't supposed to be there.

Maybe it's easiest if we start from a revision where the tests passed, even if the code needs a little cleanup.

**Comment by Jakob-Naucke** [comment]
Here's a minimal diff that at least gives me 13/15 passes:
```diff
diff --git a/crds/src/lib.rs b/crds/src/lib.rs
index 3cc38f8..1d0c614 100644
--- a/crds/src/lib.rs
+++ b/crds/src/lib.rs
@@ -19 +18,0 @@ pub struct ConfidentialClusterSpec {
-#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
@@ -33 +32 @@ pub struct Trustee {
-#[derive(CustomResource, Debug, Clone, Deserialize, Serialize, JsonSchema)]
+#[derive(CustomResource, Default, Debug, Clone, Deserialize, Serialize, JsonSchema)]
diff --git a/operator/src/trustee.rs b/operator/src/trustee.rs
index 10f75e2..18b21a1 100644
--- a/operator/src/trustee.rs
+++ b/operator/src/trustee.rs
@@ -683 +683 @@ mod tests {
-            ..Default::default()
+            spec: KbsConfigSpec::default(),
@@ -791,0 +792 @@ mod tests {
+                    metadata: ObjectMeta::default(),
@@ -797 +797,0 @@ mod tests {
-                    ..Default::default()
@@ -807 +807,4 @@ mod tests {
-                let kbs_config = KbsConfig::default();
+                let kbs_config = KbsConfig {
+                    metadata: ObjectMeta::default(),
+                    spec: KbsConfigSpec::default(),
+                };
@@ -862 +865 @@ mod tests {
-                    ..Default::default()
+                    metadata: ObjectMeta::default(),
```
Is that a good point to start or did you mean something else?

File: operator/src/trustee.rs [hunk 1]
```diff
@@ -426,3 +426,475 @@ pub async fn generate_kbs(
...
+        .unwrap();
+        let client = mock_client(StatusCode::OK, created_secret_json.into_bytes());
+
+        let result = generate_kbs_https_certificate(client, "test-ns").await;
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn test_generate_kbs_configurations_success() {
+        // This function creates three configmaps in a loop.
```
Comment (Jakob-Naucke) [inline operator/src/trustee.rs:654]: That's a true statement, but it's not very clear how it affects the test code below.

**Review by Jakob-Naucke — COMMENTED** [review]
- Can you check in your `Cargo.lock` too?
- The linter is failing because the code isn't formatted right. Can you run `cargo fmt` and update?
- Can you add a test in `.github` to run these tests in the CI?

Also tarpaulin (coverage tool) says `|| operator/src/trustee.rs: 152/169` which is pretty good :+1:

File: operator/src/trustee.rs [hunk 2]
```diff
@@ -426,3 +426,475 @@ pub async fn generate_kbs(
...
+            // An async block that returns a future
+            async move { Ok::<_, Infallible>(response) }
+        });
+
+        // Create a fake Client with our mock service
+        Client::new(mock_svc, "default")
+    }
+
+    // -----------------------------------------------------------------
+    // Category 1: Pure Logic Functions
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply operator/src/trustee.rs:472]: Done for this part. thanks

File: operator/src/trustee.rs [hunk 3]
```diff
@@ -426,3 +426,475 @@ pub async fn generate_kbs(
...
+    // -----------------------------------------------------------------
+
+    #[test]
+    fn test_get_image_pcrs_success() {
+        // 1. Prepare input data: a valid ConfigMap
+        let mut data = BTreeMap::new();
+        // THIS IS THE FIX: The JSON now includes all required fields for the `Os` struct.
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply operator/src/trustee.rs:479]: cleaned it.

File: operator/src/trustee.rs [hunk 4]
```diff
@@ -426,3 +426,475 @@ pub async fn generate_kbs(
...
+
+        let config_map = ConfigMap {
+            data: Some(data),
+            ..Default::default()
+        };
+
+        // 2. Call the function under test
+        let result = get_image_pcrs(config_map);
+
+        // 3. Assert the result
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply operator/src/trustee.rs:499]: Removed them.

File: operator/src/trustee.rs [hunk 5]
```diff
@@ -426,3 +426,475 @@ pub async fn generate_kbs(
...
+        let client = mock_client(StatusCode::OK, created_cm_json.into_bytes());
+
+        // 3. Call the function under test
+        let result = create_reference_value_config_map(client, "test-ns", "test-rv-map").await;
+
+        // 4. Assert the result: We expect the function to complete successfully without any errors.
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply operator/src/trustee.rs:574]: added.

File: operator/src/trustee.rs [hunk 6]
```diff
@@ -426,3 +426,475 @@ pub async fn generate_kbs(
...
+
+        // 3. Call the function under test
+        let result = create_reference_value_config_map(client, "test-ns", "test-rv-map").await;
+
+        // 4. Assert the result: Because the `info_if_exists!` macro catches the 409 error
+        // and treats it as non-fatal, we expect the function to still return Ok(()).
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply operator/src/trustee.rs:596]: so you mean just keep one line '#[tokio::test]' in the whole file, remove others right?

File: operator/src/trustee.rs (hunk 1 above)
Comment (6-dehan) [reply operator/src/trustee.rs:654]: Describe it a little more, please check if that works?

File: operator/src/trustee.rs [hunk 7]
```diff
@@ -426,3 +426,475 @@ pub async fn generate_kbs(
...
+    async fn test_generate_kbs_configurations_success() {
+        // This function creates three configmaps in a loop.
+        let created_cm_json = serde_json::to_string(&ConfigMap {
+            metadata: ObjectMeta {
+                name: Some("dummy-cm".to_string()),
+                ..Default::default()
+            },
+            ..Default::default()
+        })
+        .unwrap();
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply operator/src/trustee.rs:662]: okay, adjusted a little bit, please help to review again. thanks

File: operator/src/trustee.rs [hunk 8]
```diff
@@ -426,3 +426,475 @@ pub async fn generate_kbs(
...
+                    .status(StatusCode::OK)
+                    .body(Body::from(
+                        serde_json::to_string(&cm).unwrap().into_bytes(),
+                    ))
+                    .unwrap()
+            } else {
+                // For any unexpected request, return 404 Not Found
+                Response::builder()
+                    .status(StatusCode::NOT_FOUND)
+                    .body(Body::empty())
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply operator/src/trustee.rs:755]: changed. Please check.

File: operator/src/trustee.rs [hunk 9]
```diff
@@ -27,7 +27,7 @@ pub struct RvContextData {
 }
 
 /// Sync with clevis-pin-trustee::Key
-#[derive(Serialize)]
+#[derive(Serialize, serde::Deserialize)]
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply operator/src/trustee.rs:30]: updated.
